hphp/hack/src/typing/typing.ml

```diff
@@ -2806,12 +2808,35 @@ and case_list parent_locals ty env switch_pos cl dfl =
   let (env, tcl) =
     let rec case_list env = function
       | [] -> (env, [])
+      | (((_, pos, _) as e), b) :: another :: rl ->
+        let env = initialize_next_cont env in
+        let (env, te, _) = expr env e ~allow_awaitable:(*?*) false in
+        let (env, tb) = block env b in
+        let last = List.is_empty (another :: rl) && Option.is_none dfl in
+        let ((_, next_pos, _), _) = another in
+        check_fallthrough
+          env
+          switch_pos
+          pos
+          ~next_pos:(Some next_pos)
+          b
+          ~last
+          ~is_default:false;
+        let (env, tcl) = case_list env (another :: rl) in
+        (env, (te, tb) :: tcl)
       | (((_, pos, _) as e), b) :: rl ->
```

Comment:
Copy-pasting this code makes it harder to see what's going on, and how these cases differ. Instead, you can add an extra pattern match on `rl` and you don't need to repeat yourself in the `case_list` function.

```
let next_pos = match rl with
| ((_, next_pos, _), _) :: _ -> Some next_pos
| [] -> None in
```